File: android/app/src/main/kotlin/jp/golia/mailrs/accounts/MailAccount.kt
package jp.golia.mailrs.accounts

import kotlinx.serialization.Serializable

/**
 * A mailbox somewhere else, as this app holds it.
 *
 * The credential is **not** here: it lives in encrypted storage under
 * the account's id, so a row can be logged, encoded, or shown on
 * screen without carrying a password through code that has no
 * business holding one.
 */
@Serializable
data class MailAccount(
    val id: String,
    /**
     * The address, and what the server knows this person as unless
     * [login] says otherwise.
     */
    val address: String,
    /** What to call it on screen. Empty means show the address. */
    val displayName: String = "",
    /** A preset id — `gmail`, `qq` — or `custom`. */
    val provider: String = "custom",
    val imapHost: String = "",
    val imapPort: Int = 993,
    val smtpHost: String = "",
    val smtpPort: Int = 465,
    /**
     * The login name, when the server wants something other than the
     * address. Empty means the address.
     */
    val login: String = "",
    val auth: MailProvider.AuthKind = MailProvider.AuthKind.PASSWORD,
    val skipFolders: List<String> = emptyList(),
    /** Where it sits in the list. */
    val sort: Int = 0,
) {
    /** What the server is told to call this person. */
    val loginName: String get() = login.ifEmpty { address }

    /** What a person sees. */
    val title: String get() = displayName.ifEmpty { address }

    /**
     * The second line of a row, or null.
     *
     * An account with no name of its own falls back to its address on
     * the first line, so repeating the address underneath says nothing
     * and reads as a rendering fault.
     */
    val subtitle: String? get() =
        if (displayName.isEmpty() || displayName == address) null else address

    /**
     * What is wrong with this account, in words a person can act on.
     *
     * Checked here rather than at the server so a set-up screen can say
     * what is missing before spending thirty seconds finding out that a
     * blank host does not resolve.
     */
    val problem: String? get() = when {
        !address.contains('@') || address.startsWith('@') || address.endsWith('@') ->
            "That is not an email address"
        imapHost.isBlank() -> "The incoming server needs a name"
        smtpHost.isBlank() -> "The outgoing server needs a name"
        imapPort !in 1..65535 || smtpPort !in 1..65535 ->
            "A port must be between 1 and 65535"
        else -> null
    }

    companion object {
        /**
         * A stable id for an address.
         *
         * Derived rather than random so adding the same account twice
         * is the same row rather than two — and so a stored credential
         * survives a list rebuilt from scratch.
         */
        fun idFor(address: String): String {
            var h = 0xcbf29ce484222325uL
            for (c in address.lowercase()) {
                h = h xor c.code.toULong()
                h *= 0x100000001b3uL
            }
            return "acct-$h"
        }

        /** A row filled in from what is known about the address. */
        fun make(address: String, displayName: String = "", sort: Int = 0): MailAccount {
            val known = MailProvider.forAddress(address)
            val domain = address.substringAfterLast('@', "")
            val p = known ?: MailProvider.guess(domain)
            return MailAccount(
                id = idFor(address),
                address = address,
                displayName = displayName,
                provider = if (known == null) "custom" else p.label.lowercase(),
                imapHost = p.imapHost, imapPort = p.imapPort,
                smtpHost = p.smtpHost, smtpPort = p.smtpPort,
                auth = p.auth,
                skipFolders = p.skipFolders,
                sort = sort,
            )
        }

        /**
         * A colour per mailbox, so a merged list can say which is which.
         *
         * Derived from the id rather than stored: the same account is
         * the same colour on every launch, and there is nothing to keep
         * in step.
         */
        val palette = listOf(
            0xFF4285F4, 0xFF12B7F5, 0xFFEA4335, 0xFF34A853,
            0xFFA142F4, 0xFFF4B400, 0xFFFF6D00, 0xFF00897B,
        )

        fun colourFor(id: String): Long {
            var h = 0xcbf29ce484222325uL
            for (b in id.encodeToByteArray()) {
                h = h xor b.toUByte().toULong()
                h *= 0x100000001b3uL
            }
            return palette[(h % palette.size.toULong()).toInt()]
        }
    }
}

File: ios/Mailrs/Accounts/MailAccount.swift
import Foundation

/// A mailbox somewhere else, as this app holds it.
///
/// The credential is **not** here: it lives in the keychain under the
/// account's id, so a row can be logged, encoded, or shown on screen
/// without carrying a password through code that has no business
/// holding one.
/// How mail is read.
///
/// Not a preference: it changes what a mailbox *is*. POP3 has one
/// folder, no server-side flags and no stable message numbers, so read
/// state can only be kept on this device and anything filed elsewhere
/// is invisible. Saying so in the type keeps every reader honest about
/// it.
enum Incoming: String, Codable, Equatable, Sendable {
    case imap
    case pop3
    case jmap
}

struct MailAccount: Codable, Equatable, Identifiable, Sendable {
    let id: String
    /// The address, and what the server knows this person as unless
    /// `login` says otherwise.
    var address: String
    /// What to call it on screen. Empty means show the address.
    var displayName: String
    /// A preset id — `gmail`, `qq` — or `custom`.
    var provider: String
    var imapHost: String
    var imapPort: UInt16
    var smtpHost: String
    var smtpPort: UInt16
    /// The login name, when the server wants something other than the
    /// address. Empty means the address.
    var login: String
    /// How mail is read from this account.
    ///
    /// Defaulted, so a row written before this field existed still
    /// decodes — and defaulted to IMAP, which is what every account
    /// written before it existed was.
    var incoming: Incoming = .imap
    var auth: MailProvider.AuthKind
    /// Folders this account does not read.
    var skipFolders: [String]
    /// Where it sits in the list.
    var sort: Int

    /// What the server is told to call this person.
    var loginName: String { login.isEmpty ? address : login }
    /// What a person sees.
    var title: String { displayName.isEmpty ? address : displayName }

    /// A stable id for an address.
    ///
    /// Derived rather than random so adding the same account twice is
    /// the same row rather than two — and so a keychain item survives
    /// a list rebuilt from scratch.
    static func id(for address: String) -> String {
        "acct-" + address.lowercased().unicodeScalars.reduce(into: UInt64(0xcbf2_9ce4_8422_2325)) {
            h, c in
            h ^= UInt64(c.value)
            h = h &* 0x100_0000_01b3
        }.description
    }

    /// A row filled in from what is known about the address.
    static func make(address: String, displayName: String = "", sort: Int = 0) -> MailAccount {
        let known = MailProvider.forAddress(address)
        let domain = address.split(separator: "@").last.map(String.init) ?? ""
        let p = known ?? MailProvider.guess(forDomain: domain)
        return MailAccount(
            id: id(for: address),
            address: address,
            displayName: displayName,
            provider: known == nil ? "custom" : p.label.lowercased(),
            imapHost: p.imapHost, imapPort: p.imapPort,
            smtpHost: p.smtpHost, smtpPort: p.smtpPort,
            login: "",
            auth: p.auth,
            skipFolders: p.skipFolders,
            sort: sort)
    }

    /// What is wrong with this account, in words a person can act on.
    ///
    /// Checked here rather than at the server so a set-up screen can
    /// say what is missing before spending thirty seconds finding out
    /// that a blank host does not resolve.
    var problem: String? {
        if !address.contains("@") || address.hasPrefix("@") || address.hasSuffix("@") {
            return "That is not an email address"
        }
        if imapHost.trimmingCharacters(in: .whitespaces).isEmpty {
            return "The incoming server needs a name"
        }
        if imapPort == 0 { return "A port must be between 1 and 65535" }
        // JMAP submits over the same API it reads from, so there is no
        // second server and demanding one refuses an account that is
        // perfectly well described. Every other kind has both.
        if incoming == .jmap { return nil }
        if smtpHost.trimmingCharacters(in: .whitespaces).isEmpty {
            return "The outgoing server needs a name"
        }
        if smtpPort == 0 { return "A port must be between 1 and 65535" }
        return nil
    }
}
